dawn/src/dawn/Optimizer/PassRemoveScalars.cpp

```diff
@@ -0,0 +1,241 @@
+          throw std::runtime_error(dawn::format("Unsupported statement at line %d",
+                                                exprStmt->getSourceLocation().Line)); // e.g. i++;
+        }
+      }
+    }
+    // Need to treat if statements differently. There are block statements inside.
+    if(const std::shared_ptr<iir::IfStmt> ifStmt =
+           std::dynamic_pointer_cast<iir::IfStmt>(*stmtIt)) {
+
+      if(isExprScalar(ifStmt->getCondExpr(), metadata)) {
+        throw std::runtime_error(
+            dawn::format("If-condition is scalar at line %d. It is not yet supported.",
+                         ifStmt->getSourceLocation().Line)); // See reason above.
+      }
+
+      DAWN_ASSERT_MSG(ifStmt->getThenStmt()->getKind() == iir::Stmt::Kind::BlockStmt,
+                      "Then statement must be a block statement.");
+      removeScalarsFromBlockStmt(*std::dynamic_pointer_cast<iir::BlockStmt>(ifStmt->getThenStmt()),
+                                 scalarToLastRhsMap, metadata);
+
+      if(ifStmt->hasElse()) {
+        DAWN_ASSERT_MSG(ifStmt->getElseStmt()->getKind() == iir::Stmt::Kind::BlockStmt,
+                        "Else statement must be a block statement.");
+        removeScalarsFromBlockStmt(
+            *std::dynamic_pointer_cast<iir::BlockStmt>(ifStmt->getElseStmt()), scalarToLastRhsMap,
+            metadata);
+      }
+
+    } else { // Not an if statement
+      auto scalarAccessID = getScalarWrittenByStatement(*stmtIt, metadata);
+      if(scalarAccessID.has_value()) { // Writing to a scalar variable
+        scalarToLastRhsMap[*scalarAccessID] = getRhsOfAssignment(*stmtIt);
+        stmtIt = blockStmt.erase(stmtIt);
+        continue; // to next statement
+      }
+    }
+    ++stmtIt;
+  }
+}
+
+void removeScalarsFromDoMethod(iir::DoMethod& doMethod, iir::StencilMetaInformation& metadata) {
+
+  // Map from variable's access id to last rhs assigned to the variable
+  std::unordered_map<int, std::shared_ptr<const iir::Expr>> scalarToLastRhsMap;
+
+  removeScalarsFromBlockStmt(doMethod.getAST(), scalarToLastRhsMap, metadata);
+
+  // Recompute the accesses metadata of all statements (removed variables means removed
+  // accesses)
+  computeAccesses(metadata, doMethod.getAST().getStatements());
+  // Metadata cleanup
+  for(const auto& pair : scalarToLastRhsMap) {
+    int scalarAccessID = pair.first;
+    metadata.removeAccessID(scalarAccessID);
+  }
+}
+} // namespace
+
+bool PassRemoveScalars::run(
+    const std::shared_ptr<iir::StencilInstantiation>& stencilInstantiation) {
+  // TODO: sir from GTClang currently not supported: it introduces compound assignments (handling
+  // them complicates this pass). For now disabling the pass for cartesian grids.
+  // We should remove compound assignments and increment/decrement ops from SIR/IIR. They are
+  // syntactic sugar, only useful at DSL level.
+  if(stencilInstantiation->getIIR()->getGridType() == ast::GridType::Cartesian) {
```

Comment:
what do compound statements have to do with the grid type?

Reply:
Cartesian tests and clang-gridtools stencils currently contain compound statements, while unstructured tests/examples don't. As this pass is only critical for unstructured, enabling it for cartesian is deferred to a follow up task.

Reply:
Nothing prevents compound statements from being used in unstructured grids.

This just seems to be a workaround to get the tests to pass somehow and seems far from ideal

Reply:
I realize that from what I wrote it seems a workaround. But https://github.com/MeteoSwiss-APN/dawn/blob/b30672a5b26f834e852b4c7121dbe4d071a87510/dawn/src/dawn/Optimizer/PassRemoveScalars.cpp#L108-L111
actually makes compound statements illegal for unstructured. There are currently no use cases for them in our unstructured examples/tests.

Reply:
I agree that we should avoid if-statements like this in the code base as much as possible. I do feel like if the effort is reasonable we should take care to support all use-cases with new features.

Reply:
This if is a temporary solution until we do the cleanup of removing the unnecessary compound assignments from SIR.

Reply:
My intention was to introduce it incrementally, first unstructured then cartesian.
But I understand this is not what everybody wants, so the PR will have to wait.

Reply:
Having discussed this with the others, we believe that for now we can live with this temporary solution. Once a decision is taken we will change this.

Reply:
@twicki you can continue with the review, so that we can try to merge by tomorrow. I'll create the follow up task to extend the implementation to cartesian.

Reply:
why can we not move forward with fixing this dependency first and merge this once that is done? I see no strong reason to merge this is that form, or am I missing something?

Reply:
Fixing the dependency would mean writing new code either:

- in this pass to treat compound statements, increment, decrement ops as special cases, but such code would need to be deleted if we decide to remove them from iir, or
- in another pass to remove c.s., ++, --, but such pass won't be needed if we decide to remove them from sir.

The temporary solution is to not waste time in writing code that might be deleted later and to allow us to proceed in the path of fixing stage location types, which is critical for unstructured, without waiting for a decision on c.s., ++, --.

Reply:
This is now resolved, we just do not execute this if we find unsupported statements